node/src/components/fetcher/trie_fetcher.rs

```diff
@@ -0,0 +1,342 @@
+        REv: ReactorEventT<TrieOrChunkedData> + From<FetcherRequest<I, TrieOrChunkedData>>,
+    {
+        if !chunk.verify() {
+            match sender {
+                None => {
+                    error!(?chunk, "got an invalid chunk from storage");
+                    return Effects::new();
+                }
+                Some(sender) => {
+                    warn!(?sender, ?chunk, "got an invalid chunk from sender");
+                    // TODO: would be good to re-request from someone else instead of the same
+                    // node...
+                    let id = TrieOrChunkedDataId(chunk.proof().index(), chunk.proof().root_hash());
+                    return effect_builder
+                        .fetch(id, sender)
+                        .event(move |fetch_result| Event::TrieOrChunkFetched { id, fetch_result });
+                }
+            }
+        }
+        let digest = chunk.proof().root_hash();
+        let index = chunk.proof().index();
+        let count = chunk.proof().count();
+        let mut partial_chunks = match self.partial_chunks.remove(&digest) {
+            None => {
+                error!(%digest, %index, "got a chunk that wasn't requested");
+                return Effects::new();
+            }
+            Some(partial_chunks) => partial_chunks,
+        };
+
+        // Add the downloaded chunk to cache.
+        let _ = partial_chunks.chunks.insert(index, chunk);
+        // If it was downloaded from a peer, save the information.
+        partial_chunks.mutate_sender(sender);
+
+        // Check if we can now return a complete trie.
+        match partial_chunks.missing_chunk(count) {
+            Some(missing_index) => {
+                let peer = match partial_chunks.peers.last() {
+                    Some(peer) => peer.clone(),
+                    None => {
+                        debug!(
+                            %digest, %missing_index,
+                            "no peers to download the next chunk from, giving up",
+                        );
+                        return partial_chunks
+                            .responder
+                            .respond(Err(TrieFetcherError::Absent(digest, index)))
+                            .ignore();
+                    }
+                };
+                let next_id = TrieOrChunkedDataId(missing_index, digest);
+                self.try_download_chunk(effect_builder, next_id, peer, partial_chunks)
+            }
+            None => match partial_chunks.assemble_chunks(count) {
+                Ok(trie) => partial_chunks
+                    .responder
+                    .respond(into_response(Box::new(trie), partial_chunks.sender))
+                    .ignore(),
+                Err(error) => {
+                    error!(%digest, %error,
+                        "error while assembling a complete trie",
+                    );
+                    partial_chunks.responder.respond(Err(error.into())).ignore()
```

Comment:
That means cryptographically verified data failed to deserialize, right? Not sure where that's handled, but in fast sync that would probably be a reason to give up and shut down.